src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -0,0 +1,518 @@
+
+# Weirdly, it turns out that NFS is internally an alias to AWS EBS.
+
+
+class K8sStorageClient:
+    def __init__(
+        self, k8s_api: ApiClient, timeout: int, logger: BoundLogger
+    ) -> None:
+        self.k8s_api = k8s_api
+        self.api = client.CoreV1Api(k8s_api)
+        self.timeout = timeout
+        self.logger = logger
+
+    async def aclose(self) -> None:
+        await self.k8s_api.close()
+
+    def get_std_metadata(self, name: str) -> V1ObjectMeta:
+        return V1ObjectMeta(
+            name=name,
+            labels={"argocd.argoproj.io/instance": "nublado-users"},
+            annotations={
+                "argocd.argoproj.io/compare-options": "IgnoreExtraneous",
+                "argocd.argoproj.io/sync-options": "Prune=false",
+            },
+        )
+
+    async def create_user_namespace(self, namespace: str) -> None:
+        self.logger.info(f"Attempting creation of namespace '{namespace}'")
+        try:
+            await self._k8s_create_namespace(namespace)
+        except ApiException as e:
+            if e.status == 409:
+                self.logger.info(f"Namespace {namespace} already exists")
+                # ... but we know that we don't have a lab for the user,
+                # because we got this far.  So there's a stranded namespace,
+                # and we should delete it and recreate it.
+                #
+                # The spec actually calls for us to delete the lab and then the
+                # namespace, but let's just remove the namespace, which should
+                # also clean up all its contents.
+                await self.delete_namespace(namespace)
+                # Just try again, and return *that* one's return value.
+                return await self.create_user_namespace(namespace)
+            else:
+                estr = f"Failed to create namespace {namespace}: {e}"
+                self.logger.exception(estr)
+                raise NSCreationError(estr)
+        # We don't want to give control back until the namespace exists,
+        # because the next step will be creating objects in that namespace.
+        await self._wait_for_operation(
+            namespace=namespace,
+            item="namespace",
+            operation=ObjectOperation.CREATION,
+            interval=0.2,
+        )
+
+    async def _k8s_create_namespace(self, ns_name: str) -> None:
+        await asyncio.wait_for(
+            self.api.create_namespace(
+                V1Namespace(metadata=self.get_std_metadata(name=ns_name))
+            ),
+            self.timeout,
+        )
+
+    async def create_secrets(
+        self,
+        secret_list: List[LabSecret],
+        username: str,
+        token: str,
+        source_ns: str,
+        target_ns: str,
+    ) -> None:
+        pull_secrets = [
+            x for x in secret_list if x.secret_name == "pull_secret"
+        ]
+        secrets = [x for x in secret_list if x.secret_name != "pull_secret"]
```

Comment:
The storage layer shouldn't be special-casing things based on their names (that's business logic). The caller should be responsible for passing in a list of secrets and a separate pull secret if they are to be handled differently, or even better, create the pull secret in a separate method from the merged secret.